pkg/credentialprovider/plugin/framework/plugin.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"k8s.io/kubernetes/pkg/credentialprovider/apis/credentials/install"
+	"k8s.io/kubernetes/pkg/credentialprovider/apis/credentials/v1alpha1"
+)
+
+var (
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+)
+
+func init() {
+	install.Install(scheme)
+}
+
+// CredentialProvider is an interface implemented by the kubelet credential provider plugin to fetch
+// the username/password based on the provided image name.
+type CredentialProvider interface {
+	GetCredentials(ctx context.Context, image string, args []string) (response *v1alpha1.CredentialProviderResponse, err error)
+}
+
+// ExecPlugin implements the exec-based plugin for fetching credentials that is invoked by the kubelet.
+type ExecPlugin struct {
+	plugin CredentialProvider
+}
+
+// NewCredentialProvider returns an instance of execPlugin that fetches
+// credentials based on the provided plugin implementing the CredentialProvider interface.
+func NewCredentialProvider(plugin CredentialProvider) *ExecPlugin {
+	return &ExecPlugin{plugin}
+}
+
+// Run executes the credential provider plugin. Required information for the plugin request (in
+// the form of v1alpha1.CredentialProviderRequest) is provided via stdin from the kubelet.
+// The CredentialProviderResponse, containing the username/password required for pulling
+// the provided image, will be sent back to the kubelet via stdout.
+func (e *ExecPlugin) Run(ctx context.Context) error {
```

Comment:
what calls Run? if Run returns an error, should that always be echoed to stderr with a non-zero return code? If so, suggest a RunOrDie helper that does that, and update the examples to call that (should make their main() methods a one-liner)

Reply:
Resolving since I removed the framework package which mostly served as a reference implementation.